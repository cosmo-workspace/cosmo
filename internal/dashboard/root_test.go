/*
Copyright © 2023 NAME HERE cosmo-workspace
*/
package dashboard

import (
	"bytes"
	"io"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
)

func TestNewRootCmd(t *testing.T) {

	tests := []struct {
		name string
		args []string
	}{
		{
			name: "❌ flags are required",
			args: []string{
				"--ldap-url=xxxx",
			},
		},
		{
			name: "❌ cookie-hashkey is minimum 16 characters",
			args: []string{
				"--cookie-hashkey=123456789012345",
				"--cookie-blockkey=1234567890123456",
				"--insecure",
			},
		},
		{
			name: "❌ cookie-blockkey is minimum 16 characters",
			args: []string{
				"--cookie-hashkey=1234567890123456",
				"--cookie-blockkey=yyy",
				"--insecure",
			},
		},
		{
			name: "❌ ldap-url is invalid",
			args: []string{
				"--cookie-hashkey=1234567890123456",
				"--cookie-blockkey=1234567890123456",
				"--ldap-url=ldap://sss  ssss",
				"--ldap-basedn=xxxx",
				"--ldap-user-attr=yyyy",
				"--insecure",
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			outBuf := bytes.NewBufferString("")
			o := &options{}
			rootCmd := NewRootCmd(o)
			rootCmd.SetOutput(outBuf)
			rootCmd.SetArgs(tt.args)
			_ = rootCmd.Execute()
			out, _ := io.ReadAll(outBuf)
			snaps.MatchSnapshot(t, string(out))
		})
	}
}
