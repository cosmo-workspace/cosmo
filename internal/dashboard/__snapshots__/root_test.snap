
[TestNewRootCmd/❌_flags_are_required - 1]
Error: required flag(s) "cookie-blockkey", "cookie-hashkey", "tls-cert", "tls-key" not set
Usage:
  dashboard [flags]

Flags:
      --add_dir_header                    If true, adds the file directory to the header of the log messages
      --alsologtostderr                   log to standard error as well as files (no effect when -logtostderr=true)
      --cookie-blockkey string            Cookie blockkey
      --cookie-domain string              Cookie domain name
      --cookie-hashkey string             Cookie hashkey
      --cookie-session-name string        Cookie session name (default "cosmo-auth")
      --graceful-shutdown-seconds int     Graceful shutdown seconds (default 10)
  -h, --help                              help for dashboard
      --insecure                          start http server not https server
      --kubeconfig string                 Paths to a kubeconfig. Only required if out-of-cluster.
      --ldap-binddn string                [bind mode] ex: cn=%s,ou=users,dc=example,dc=com  '%s' is replaced by the userid.
      --ldap-ca-cert string               ca cert file path
      --ldap-insecure-skip-verify         Skip server certificate chain and hostname validation
      --ldap-search-basedn string         [search mode] ex: dc=example,dc=com
      --ldap-search-binddn string         [search mode] ex: cn=admin,dc=example,dc=com '%s' is replaced by the userid.
      --ldap-search-filter string         [search mode] ex: (uid=%s)  '%s' is replaced by the userid.
      --ldap-search-password string       [search mode] password for search bindDN.
      --ldap-start-tls                    Enables StartTLS functionality
      --ldap-url string                   LDAP URL. ldap[s]://hostname.or.ip[:port]
      --log_backtrace_at traceLocation    when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                    If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                   If non-empty, use this log file (no effect when -logtostderr=true)
      --log_file_max_size uint            Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                       log to standard error instead of files (default true)
      --maxage-minutes int                session maxage minutes (default 720)
      --one_output                        If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --port int                          Port for dashboard server (default 8443)
      --serve-dir string                  Static file dir to serve (default "/app/public")
      --skip_headers                      If true, avoid header prefixes in the log messages
      --skip_log_headers                  If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity          logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
      --timeout-seconds int               Timeout seconds for response (default 3)
      --tls-cert string                   TLS certificate file path (default "tls.crt")
      --tls-key string                    TLS key file path (default "tls.key")
  -v, --v Level                           number for the log level verbosity
      --version                           version for dashboard
      --vmodule moduleSpec                comma-separated list of pattern=N settings for file-filtered logging
      --zap-devel                         Development Mode defaults(encoder=consoleEncoder,logLevel=Debug,stackTraceLevel=Warn). Production Mode defaults(encoder=jsonEncoder,logLevel=Info,stackTraceLevel=Error)
      --zap-encoder encoder               Zap log encoding (one of 'json' or 'console')
      --zap-log-level level               Zap Level to configure the verbosity of logging. Can be one of 'debug', 'info', 'error', or any integer value > 0 which corresponds to custom debug levels of increasing verbosity
      --zap-stacktrace-level level        Zap Level at and above which stacktraces are captured (one of 'info', 'error', 'panic').
      --zap-time-encoding time-encoding   Zap time encoding (one of 'epoch', 'millis', 'nano', 'iso8601', 'rfc3339' or 'rfc3339nano'). Defaults to 'epoch'.


---

[TestNewRootCmd/❌_cookie-hashkey_is_minimum_16_characters - 1]
Error: validation error: cookie-hashkey is minimum 16 characters
Usage:
  dashboard [flags]

Flags:
      --add_dir_header                    If true, adds the file directory to the header of the log messages
      --alsologtostderr                   log to standard error as well as files (no effect when -logtostderr=true)
      --cookie-blockkey string            Cookie blockkey
      --cookie-domain string              Cookie domain name
      --cookie-hashkey string             Cookie hashkey
      --cookie-session-name string        Cookie session name (default "cosmo-auth")
      --graceful-shutdown-seconds int     Graceful shutdown seconds (default 10)
  -h, --help                              help for dashboard
      --insecure                          start http server not https server
      --kubeconfig string                 Paths to a kubeconfig. Only required if out-of-cluster.
      --ldap-binddn string                [bind mode] ex: cn=%s,ou=users,dc=example,dc=com  '%s' is replaced by the userid.
      --ldap-ca-cert string               ca cert file path
      --ldap-insecure-skip-verify         Skip server certificate chain and hostname validation
      --ldap-search-basedn string         [search mode] ex: dc=example,dc=com
      --ldap-search-binddn string         [search mode] ex: cn=admin,dc=example,dc=com '%s' is replaced by the userid.
      --ldap-search-filter string         [search mode] ex: (uid=%s)  '%s' is replaced by the userid.
      --ldap-search-password string       [search mode] password for search bindDN.
      --ldap-start-tls                    Enables StartTLS functionality
      --ldap-url string                   LDAP URL. ldap[s]://hostname.or.ip[:port]
      --log_backtrace_at traceLocation    when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                    If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                   If non-empty, use this log file (no effect when -logtostderr=true)
      --log_file_max_size uint            Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                       log to standard error instead of files (default true)
      --maxage-minutes int                session maxage minutes (default 720)
      --one_output                        If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --port int                          Port for dashboard server (default 8443)
      --serve-dir string                  Static file dir to serve (default "/app/public")
      --skip_headers                      If true, avoid header prefixes in the log messages
      --skip_log_headers                  If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity          logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
      --timeout-seconds int               Timeout seconds for response (default 3)
      --tls-cert string                   TLS certificate file path (default "tls.crt")
      --tls-key string                    TLS key file path (default "tls.key")
  -v, --v Level                           number for the log level verbosity
      --version                           version for dashboard
      --vmodule moduleSpec                comma-separated list of pattern=N settings for file-filtered logging
      --zap-devel                         Development Mode defaults(encoder=consoleEncoder,logLevel=Debug,stackTraceLevel=Warn). Production Mode defaults(encoder=jsonEncoder,logLevel=Info,stackTraceLevel=Error)
      --zap-encoder encoder               Zap log encoding (one of 'json' or 'console')
      --zap-log-level level               Zap Level to configure the verbosity of logging. Can be one of 'debug', 'info', 'error', or any integer value > 0 which corresponds to custom debug levels of increasing verbosity
      --zap-stacktrace-level level        Zap Level at and above which stacktraces are captured (one of 'info', 'error', 'panic').
      --zap-time-encoding time-encoding   Zap time encoding (one of 'epoch', 'millis', 'nano', 'iso8601', 'rfc3339' or 'rfc3339nano'). Defaults to 'epoch'.


---

[TestNewRootCmd/❌_cookie-blockkey_is_minimum_16_characters - 1]
Error: validation error: cookie-blockkey is minimum 16 characters
Usage:
  dashboard [flags]

Flags:
      --add_dir_header                    If true, adds the file directory to the header of the log messages
      --alsologtostderr                   log to standard error as well as files (no effect when -logtostderr=true)
      --cookie-blockkey string            Cookie blockkey
      --cookie-domain string              Cookie domain name
      --cookie-hashkey string             Cookie hashkey
      --cookie-session-name string        Cookie session name (default "cosmo-auth")
      --graceful-shutdown-seconds int     Graceful shutdown seconds (default 10)
  -h, --help                              help for dashboard
      --insecure                          start http server not https server
      --kubeconfig string                 Paths to a kubeconfig. Only required if out-of-cluster.
      --ldap-binddn string                [bind mode] ex: cn=%s,ou=users,dc=example,dc=com  '%s' is replaced by the userid.
      --ldap-ca-cert string               ca cert file path
      --ldap-insecure-skip-verify         Skip server certificate chain and hostname validation
      --ldap-search-basedn string         [search mode] ex: dc=example,dc=com
      --ldap-search-binddn string         [search mode] ex: cn=admin,dc=example,dc=com '%s' is replaced by the userid.
      --ldap-search-filter string         [search mode] ex: (uid=%s)  '%s' is replaced by the userid.
      --ldap-search-password string       [search mode] password for search bindDN.
      --ldap-start-tls                    Enables StartTLS functionality
      --ldap-url string                   LDAP URL. ldap[s]://hostname.or.ip[:port]
      --log_backtrace_at traceLocation    when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                    If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                   If non-empty, use this log file (no effect when -logtostderr=true)
      --log_file_max_size uint            Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                       log to standard error instead of files (default true)
      --maxage-minutes int                session maxage minutes (default 720)
      --one_output                        If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --port int                          Port for dashboard server (default 8443)
      --serve-dir string                  Static file dir to serve (default "/app/public")
      --skip_headers                      If true, avoid header prefixes in the log messages
      --skip_log_headers                  If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity          logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
      --timeout-seconds int               Timeout seconds for response (default 3)
      --tls-cert string                   TLS certificate file path (default "tls.crt")
      --tls-key string                    TLS key file path (default "tls.key")
  -v, --v Level                           number for the log level verbosity
      --version                           version for dashboard
      --vmodule moduleSpec                comma-separated list of pattern=N settings for file-filtered logging
      --zap-devel                         Development Mode defaults(encoder=consoleEncoder,logLevel=Debug,stackTraceLevel=Warn). Production Mode defaults(encoder=jsonEncoder,logLevel=Info,stackTraceLevel=Error)
      --zap-encoder encoder               Zap log encoding (one of 'json' or 'console')
      --zap-log-level level               Zap Level to configure the verbosity of logging. Can be one of 'debug', 'info', 'error', or any integer value > 0 which corresponds to custom debug levels of increasing verbosity
      --zap-stacktrace-level level        Zap Level at and above which stacktraces are captured (one of 'info', 'error', 'panic').
      --zap-time-encoding time-encoding   Zap time encoding (one of 'epoch', 'millis', 'nano', 'iso8601', 'rfc3339' or 'rfc3339nano'). Defaults to 'epoch'.


---

[TestNewRootCmd/❌_ldap-url_is_invalid - 1]
Error: unknown flag: --ldap-basedn
Usage:
  dashboard [flags]

Flags:
      --add_dir_header                    If true, adds the file directory to the header of the log messages
      --alsologtostderr                   log to standard error as well as files (no effect when -logtostderr=true)
      --cookie-blockkey string            Cookie blockkey
      --cookie-domain string              Cookie domain name
      --cookie-hashkey string             Cookie hashkey
      --cookie-session-name string        Cookie session name (default "cosmo-auth")
      --graceful-shutdown-seconds int     Graceful shutdown seconds (default 10)
  -h, --help                              help for dashboard
      --insecure                          start http server not https server
      --kubeconfig string                 Paths to a kubeconfig. Only required if out-of-cluster.
      --ldap-binddn string                [bind mode] ex: cn=%s,ou=users,dc=example,dc=com  '%s' is replaced by the userid.
      --ldap-ca-cert string               ca cert file path
      --ldap-insecure-skip-verify         Skip server certificate chain and hostname validation
      --ldap-search-basedn string         [search mode] ex: dc=example,dc=com
      --ldap-search-binddn string         [search mode] ex: cn=admin,dc=example,dc=com '%s' is replaced by the userid.
      --ldap-search-filter string         [search mode] ex: (uid=%s)  '%s' is replaced by the userid.
      --ldap-search-password string       [search mode] password for search bindDN.
      --ldap-start-tls                    Enables StartTLS functionality
      --ldap-url string                   LDAP URL. ldap[s]://hostname.or.ip[:port]
      --log_backtrace_at traceLocation    when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                    If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                   If non-empty, use this log file (no effect when -logtostderr=true)
      --log_file_max_size uint            Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                       log to standard error instead of files (default true)
      --maxage-minutes int                session maxage minutes (default 720)
      --one_output                        If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --port int                          Port for dashboard server (default 8443)
      --serve-dir string                  Static file dir to serve (default "/app/public")
      --skip_headers                      If true, avoid header prefixes in the log messages
      --skip_log_headers                  If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity          logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
      --timeout-seconds int               Timeout seconds for response (default 3)
      --tls-cert string                   TLS certificate file path (default "tls.crt")
      --tls-key string                    TLS key file path (default "tls.key")
  -v, --v Level                           number for the log level verbosity
      --version                           version for dashboard
      --vmodule moduleSpec                comma-separated list of pattern=N settings for file-filtered logging
      --zap-devel                         Development Mode defaults(encoder=consoleEncoder,logLevel=Debug,stackTraceLevel=Warn). Production Mode defaults(encoder=jsonEncoder,logLevel=Info,stackTraceLevel=Error)
      --zap-encoder encoder               Zap log encoding (one of 'json' or 'console')
      --zap-log-level level               Zap Level to configure the verbosity of logging. Can be one of 'debug', 'info', 'error', or any integer value > 0 which corresponds to custom debug levels of increasing verbosity
      --zap-stacktrace-level level        Zap Level at and above which stacktraces are captured (one of 'info', 'error', 'panic').
      --zap-time-encoding time-encoding   Zap time encoding (one of 'epoch', 'millis', 'nano', 'iso8601', 'rfc3339' or 'rfc3339nano'). Defaults to 'epoch'.


---
